src/HTSPDemuxer.cpp

```diff
@@ -630,21 +654,127 @@ void CHTSPDemuxer::ParseSubscriptionSpeed ( htsmsg_t *m )
 
 void CHTSPDemuxer::ParseSubscriptionStatus ( htsmsg_t *m )
 {
-  const char *status;
+  const char *status, *error;
   status = htsmsg_get_str(m, "status");
+  error  = htsmsg_get_str(m, "subscriptionError");
 
-  // not for preTuning subscriptions
-  if (m_subscription.weight != SUBSCRIPTION_WEIGHT_DEFAULT)
+  // not for preTuning subscriptions, tag them as unknown
+  if (m_subscription.weight < SUBSCRIPTION_WEIGHT_DEFAULT)
+  {
+    m_subscription.state = SUBSCRIPTION_UNKNOWN;
     return;
+  }
 
   // this field is absent when everything is fine
   if (status != NULL)
-  {
     tvhinfo("Bad subscription status: %s", status);
-    XBMC->QueueNotification(QUEUE_INFO, status);
+
+  /* 'subscriptionError' was added in htsp v20 */
+  /* Use 'status' for older tvheadend backends */
+  if (m_conn.GetProtocol() >= 20)
+  {
+    if (error != NULL)
+    {
+      if (!strcmp("badSignal", error))
+        m_subscription.state = SUBSCRIPTION_NOSIGNAL;
+      else if (!strcmp("scrambled", error))
+        m_subscription.state = SUBSCRIPTION_SCRAMBLED;
+      else if (!strcmp("userLimit", error))
+        m_subscription.state = SUBSCRIPTION_USERLIMIT;
+      else if (!strcmp("noFreeAdapter", error))
```

Comment:
Didn't we have some time ago a discussion regarding parsing error strings? My mind has not changed since then. Is it guaranteed that these strings are part of tvheadend's API and that nobody will at any time change these strings or that at one day localized strings will appear here?